### PR TITLE
#163153870 Feature update user details

### PR DESCRIPTION
**Pull request by habibaudu**
#### What does this PR do?
- Update User Details
#### Description of Task to be completed?
- Update user details
- Write the test for user detail controller
- Validate update input
- Write the test for the validation
#### How should this be manually tested?
- Fire up postman
- log in to get a token
- Add token with Key Authorization
- URL:[here]( http://localhost/api/v1/user)

#### What are the relevant pivotal tracker stories?
[here](https://www.pivotaltracker.com/story/show/163153870)

### CODE REVIEW

File: routes/userRoutes.js [hunk 1]
```diff
@@ -12,6 +19,9 @@ const router = express.Router();
 router.route('/user')
   .all(verifyToken)
   .get(getUser)
-  .put(updateUser);
+  .patch(passwordValidator, isUsernameValidator, usernameExistsValidator,
```
Comment (oreoluwade) [inline routes/userRoutes.js]: step these middlewares to new lines like the ones on lines 23, 24, 25

Comment (habibaudu) [reply routes/userRoutes.js]: ok

File: test/user.test.js [hunk 2]
```diff
@@ -401,4 +401,169 @@ describe('USER TEST SUITE', () => {
         expect(response.body.error).to.equal('Invalid token');
       });
   });
+
+  describe('User Update Validations', () => {
+    it('should return No token provided, when No token is provided',
```
Comment (oreoluwade) [inline test/user.test.js]: Please change the wording of this spec. It is better written as:

`Should not allow update when no token is provided`

You don't need to write the exact error message in the spec description.

File: test/user.test.js [hunk 3]
```diff
@@ -401,4 +401,169 @@ describe('USER TEST SUITE', () => {
...
+      async () => {
+        const response = await chai.request(server)
+          .patch('/api/v1/user');
+        expect(response.status).to.equal(400);
+        expect(response.body.error).to.equal('No token provided');
+      });
+
+
+    it('should return Invalid token, when token is Invalid',
```
Comment (oreoluwade) [inline test/user.test.js]: `Should not allow update when the supplied token is invalid`

File: test/user.test.js [hunk 4]
```diff
@@ -401,4 +401,169 @@ describe('USER TEST SUITE', () => {
...
+      async () => {
+        const response = await chai.request(server)
+          .patch('/api/v1/user')
+          .set('Authorization', 'invalidtoken');
+        expect(response.status).to.equal(401);
+        expect(response.body.error).to.equal('Invalid token');
+      });
+
+
+    it('should return User Updated Successfully,with correct parameters',
```
Comment (oreoluwade) [inline test/user.test.js]: `Should successfully update the user when accurate parameters are supplied`

File: test/user.test.js (hunk 2 above)
Comment (habibaudu) [reply test/user.test.js]: ok

File: test/user.test.js (hunk 3 above)
Comment (habibaudu) [reply test/user.test.js]: ok will do

File: test/user.test.js [hunk 5]
```diff
@@ -401,4 +401,169 @@ describe('USER TEST SUITE', () => {
...
+          .send(newUser2)
+          .set('Authorization', firstUserToken);
+        expect(response.body.message).to.equal('User Updated Successfully');
+        expect(response.body.username).to.equal('talktoat');
+        expect(response.body.email).to.equal('talkto@gmail.com');
+        expect(response.body.isMentor).to.equal(false);
+      });
+
+
+    it('should return username already exist,for existing Username',
```
Comment (oreoluwade) [inline test/user.test.js]: `Should not allow update when the supplied username already exists `

File: test/user.test.js (hunk 4 above)
Comment (habibaudu) [reply test/user.test.js]: ok

File: test/user.test.js (hunk 5 above)
Comment (habibaudu) [reply test/user.test.js]: ok... will do

File: test/user.test.js [hunk 6]
```diff
@@ -401,4 +401,169 @@ describe('USER TEST SUITE', () => {
...
+        };
+        const response = await chai.request(server)
+          .patch('/api/v1/user')
+          .send(newUser2)
+          .set('Authorization', firstUserToken);
+        expect(response.body.error).to.equal('Username already in use');
+      });
+
+
+    it('should return username must be at least 6 ,when charcters less than 6',
```
Comment (oreoluwade) [inline test/user.test.js]: `Should not allow update when the username is not the permitted username format`

File: test/user.test.js [hunk 7]
```diff
@@ -401,4 +401,169 @@ describe('USER TEST SUITE', () => {
...
+        const response = await chai.request(server)
+          .patch('/api/v1/user')
+          .send(newUser2)
+          .set('Authorization', firstUserToken);
+        expect(response.body.error).to.equal(
+          'Your username must be at least 6 characters');
+      });
+
+
+    it('username must contain only alphabets and numbers',
```
Comment (oreoluwade) [inline test/user.test.js]: This test spec may be redundant

**Review by oreoluwade — APPROVED** [review]

**Pull request merged**